### PR TITLE
Database

### PR DESCRIPTION
**Pull request by Aditya-NeuralNetNinja**
data, src folders created

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    cnx = mysql.connect(user=user,
+                        password=os.getenv('PASSWORD'),
+                        host=host,
+                        db=db)
+    cur = cnx.cursor()
+    return cnx, cur
+
+# STEP 2 - Create db
+def build_db(db:str) -> None:
```
Comment (mitulds) [inline src/utils.py:33]: pass `cur` to the args

File: src/utils.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+def get_data(file_path:str) -> DataFrame:
+    """_summary_
+
+    Args:
+        file_path (str): _description_
+
+    Returns:
+        DataFrame: _description_
+    """
+    df = pd.read_csv(file_path)
```
Comment (mitulds) [inline src/utils.py:52]: add this line of code
`df.drop(['Unnamed: 0'], axis=1, inplace=True, errors='ignore')`

File: src/utils.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return df
+
+# STEP 4 - Build Schema/Table --- how to connect table to df???
+def build_table(db:str, table:str, col1:str, col2:str) -> None:
```
Comment (mitulds) [inline src/utils.py:56]: missing `cur` in the args

File: src/utils.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+# STEP 2 - Create db
+def build_db(db:str) -> None:
+    """Build db
+
+    Args:
+        db (str): Input db name
+    """
+    cur.execute(f'DROP DATABASE IF EXISTS {db};')
+    cur.execute(f'CREATE DATABASE {db};')
```
Comment (mitulds) [inline src/utils.py:40]: include 
```
cur.execute('SHOW DATABASES')
databases = cursor.fetchall()
return databases
```

File: src/utils.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        db (str): Database name
+        table (str): Schema name
+        col1 (str): Name of column 1
+        datatype1 (str): Column 1 datatype
+        col2 (str): Name of column 2
+        datatype2 (str): Column 2 datatype
+    """
+    cur.execute(f'USE {db};')
+    cur.execute(f'DROP TABLE IF EXISTS {table};')
+    cur.execute(f'CREATE TABLE {table}({col1} {df[col1].dtype}, {df[col2].dtype});')
```
Comment (mitulds) [inline src/utils.py:69]: modify line 69 with
`cur.execute(f'CREATE TABLE {table} ({result});')`

File: src/code.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    elif df[col_name].dtype == 'float64':
+        type = 'FLOAT'
+    
+    column_dtypes.append(f"{col_name} {type}")
+
+result = f"({', '.join(column_dtypes)})"
+print(result)
+
+placeholders = ', '.join(len(df.columns)*['%s'])
+print(placeholders)
```
Comment (mitulds) [inline src/code.py:130]: store this in a new function and in utils file. This function should take input as `df` and should `return result, placeholders`

File: src/utils.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+build_table(db='d2p1',
+            table='tutorial',
+            col1='id',
+            datatype1='int',
+            col2='course_name',
+            datatype2='varchar(20)')
```
Comment (mitulds) [inline src/utils.py:76]: remove this code from here

File: src/utils.py (hunk 5 above)
Comment (mitulds) [reply src/utils.py:69]: remove `col1` and `col2` args, and replace with `result`

**Review by Aditya-NeuralNetNinja — COMMENTED** [review]
Changes done, pushed to py2sql branch

File: src/utils.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        placeholders (str): _description_
+    """
+    total=0
+    for _,row in df.iterrows():
+        sql = f"INSERT INTO {table} VALUES({placeholders})"
+        val = tuple(row)
+        cur.execute(sql, val)
+        if cur.rowcount==1:
+            total+=1
+    cnx.commit()
```
Comment (mitulds) [inline src/utils.py:123]: add `return total`

File: src/database.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+parser = argparse.ArgumentParser(
+    prog= 'Database file',
+    description= 'Connect to mysqlserver, creates db if non-existent, build schema & table, thereby feed data to table',
+    epilog= 'https://docs.python.org/3/howto/argparse.html'
+)
+
+# Defining script args
+db_exists=False
+db_name='sales'
+file='/Users/zomato/Desktop/d2p_proj/data/supermarket_sales.csv'
```
Comment (mitulds) [inline src/database.py:15]: remove these

File: src/database.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+# Parse script arguments
+parser.add_argument("-v","--verbose", 
+                    help="increase output verbosity",
+                    action="store_true")
```
Comment (mitulds) [inline src/database.py:20]: dont need this

File: src/database.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+# STEP 2 - Create/Utilize db
+if args.db_exists:   
+    build_db(cur=cur, db=args.db_name)
+    print(db_name)
+
+else:
+# STEP 3 - Read CSV data
+    df = get_data(file_path=args.file)
+    table_name = os.path.basename(args.file).split('.')[0]
```
Comment (mitulds) [inline src/database.py:50]: should be `args.file_path`

File: src/utils.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+# STEP 7 - Close cursor, connection
+def close_all(cur:mysql.cursor.MySQLCursor, cnx:mysql.MySQLConnection) -> None:
+    """
+    Args:
+        cur (str): cursor
+        cnx (str): connection
+    """
+    cur.close()
+    cnx.close() # type: ignore
```
Comment (mitulds) [inline src/utils.py:133]: remove this

File: src/database.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    schema, placeholders = build_schema(df=df)
+
+# STEP 5 - Build Table
+    build_table(cur=cur, db=args.db_name, table=table_name, schema=schema)
+
+# STEP 6 - Insert Data into Table
+    ingest_data(cur=cur, cnx=cnx, df=df, table=table_name, placeholders=placeholders)
+        
+# STEP 7 - Close cursor, connection
+    close_all(cur=cur, cnx=cnx)
```
Comment (mitulds) [inline src/database.py:62]:
```
cur.close()
cnx.close()
```

File: src/database.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+                    help='Location of file in directory')
+
+args = parser.parse_args()
+
+# STEP 1 - connect to Mysql server
+cnx,cur = connector(user='root', host='localhost')
+
+# STEP 2 - Create/Utilize db
+if args.db_exists:   
+    build_db(cur=cur, db=args.db_name)
```
Comment (mitulds) [inline src/database.py:44]:
```
db = build_db(cur=cur, db=args.db_name)
print(db)
```

File: src/database.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    table_name = os.path.basename(args.file).split('.')[0]
+
+# STEP 4 - Build Schema
+    schema, placeholders = build_schema(df=df)
+
+# STEP 5 - Build Table
+    build_table(cur=cur, db=args.db_name, table=table_name, schema=schema)
+
+# STEP 6 - Insert Data into Table
+    ingest_data(cur=cur, cnx=cnx, df=df, table=table_name, placeholders=placeholders)
```
Comment (mitulds) [inline src/database.py:59]: assign this to a variable and then add print - line 59

**Review by Aditya-NeuralNetNinja — COMMENTED** [review]
Necessary changes done

File: src/database.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    description= 'Connect to mysqlserver, creates db if non-existent, build schema & table, thereby feed data to table',
+    epilog= 'https://docs.python.org/3/howto/argparse.html'
+)
+
+# Defining script args
+'''
+db_exists=False
+db_name='sales'
+file='/Users/zomato/Desktop/d2p_proj/data/supermarket_sales.csv'
+'''
```
Comment (mitulds) [inline src/database.py:19]: delete these

**Comment by mitulds** [comment]
you havent done all the changes that I've requested yet. I'd suggest you to go these again.

**Review by mitul23 — APPROVED** [review]
Good job!

**Review by mitulds — APPROVED** [review]

**Pull request merged**